### PR TITLE
GH-90699: Intern statically allocated strings

### PR DESCRIPTION
**Pull request by kumaraditya303**
This is similar to how strings are interned for deepfreeze.

* Issue: gh-90699

Automerge-Triggered-By: GH:ericsnowcurrently

### CODE REVIEW

File: Tools/scripts/generate_global_objects.py
Comment (arhadthedev) [inline Tools/scripts/generate_global_objects.py]: I think we can introduce a macro to half the generated function from ~600 to ~300 lines:

```suggestion
            printer.write('#define INTERN(id) \\')
            printer.write('    string = &_Py_ID(id); \\'
            printer.write('    PyUnicode_InternInPlace(&string);')
            for i in sorted(identifiers):
                printer.write(f'INTERN({i});')
            printer.write(f'#undef INTERN')
```

Comment (kumaraditya303) [reply Tools/scripts/generate_global_objects.py]: Well hiding function call in a macro makes debugging harder and has hidden control flow, so I'll pass it out for now. Thanks!

Comment (arhadthedev) [reply Tools/scripts/generate_global_objects.py]: > makes debugging harder and has hidden control flow

Initially I missed this aspect, thank you for pointing out.

**Comment by AA-Turner** [comment]
Should `pycore_runtime_init.h` be marked as generated?

https://github.com/python/cpython/blob/243ed5439c32e8517aa745bc2ca9774d99c99d0f/.gitattributes#L69-L71

A

**Comment by kumaraditya303** [comment]
> Should `pycore_runtime_init.h` be marked as generated?

That file also contains handwritten struct initialization macros so I don't think so.

**Comment by ericsnowcurrently** [comment]
Would we also move the generated parts of pycore_global_strings.h to a separate header file (for the same reasons)?

**Comment by AA-Turner** [comment]
With this PR, ~93% of the lines in `pycore_runtime_init.h` are autogenerated -- I did miss that it starts with ~120 lines of handwritten code, but for future maintenence / ease of review I'd agree with Eric that it might be better to move to its own file. (A large part is also my unfamiliarity with this part of the code, to be fair -- if it causes more hassle than it's worth then the status quo wins).

A

**Comment by kumaraditya303** [comment]
I can work on a PR to separate the generated and handwritten code but it will be a separate PR. I have other changes enqueued for static strings which depend on this PR so I want to get this in first.

File: Tools/scripts/generate_global_objects.py
Comment (ericsnowcurrently) [inline Tools/scripts/generate_global_objects.py]: Consider splitting this line for readability:

```suggestion
        printer.write("static inline void")
        with printer.block("_Py_StaticStrings_Intern(void)"):
```

or don't split the line:

```suggestion
        with printer.block("static inline void _Py_StaticStrings_Intern(void)"):
```

File: Include/internal/pycore_runtime_init.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_runtime_init.h]: Presumably this will move over to pycore_runtime_generated.h once #94051 has landed.

File: Include/internal/pycore_runtime_init.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_runtime_init.h]: It would be worth having a note here indicating that these uses of `_Py_ID()` won't be picked up by Tools/scripts/generate_global_objects.py.

File: Python/pylifecycle.c
Comment (ericsnowcurrently) [inline Python/pylifecycle.c]: This should go in `_PyUnicode_InitGlobalObjects()` (which is called in `pycore_init_global_objects()`).

File: Include/internal/pycore_runtime_init.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_runtime_init.h]: Let's stick with convention for the function name, e.g. "_PyUnicode_InitStaticStrings()".

File: Include/internal/pycore_runtime_init.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_runtime_init.h]: Multiple lines per entry makes this a little harder to read, but that probably doesn't matter too much for generated code.  It might be a little less efficient to store into a variable but it probably isn't significant and I expect the C compiler optimizes that away.  So there probably isn't anything to change here.

That said, it would be nice if we didn't need an intermediate variable for this.  This would require a friendlier version of `PyUnicode_InternInPlace()`, which is outside the scope of this PR.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
This mostly looks good.  I just have a few comments.

File: Tools/scripts/generate_global_objects.py
Comment (kumaraditya303) [reply Tools/scripts/generate_global_objects.py]: done

File: Include/internal/pycore_runtime_init.h
Comment (kumaraditya303) [reply Include/internal/pycore_runtime_init.h]: done

File: Python/pylifecycle.c
Comment (kumaraditya303) [reply Python/pylifecycle.c]: done

File: Include/internal/pycore_runtime_init.h
Comment (kumaraditya303) [reply Include/internal/pycore_runtime_init.h]: Added

File: Include/internal/pycore_runtime_init.h
Comment (kumaraditya303) [reply Include/internal/pycore_runtime_init.h]: Yes

File: Tools/scripts/generate_global_objects.py
Comment (ericsnowcurrently) [inline Tools/scripts/generate_global_objects.py]: The comment helps.  However, I recommend moving it down to the generated `_Py_ID()` in generate_runtime_init(), along with a slight adjustment.
```suggestion
```

File: Tools/scripts/generate_global_objects.py
Comment (ericsnowcurrently) [inline Tools/scripts/generate_global_objects.py:294]: As noted above, the earlier comment might be more helpful here.  Also, with a slight adjustment:

```suggestion
                # This use of _Py_ID() is ignored by iter_global_strings()
                # since iter_files() ignores .h files.
                printer.write(f'string = &_Py_ID({i});')
```

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Thanks for make those updates!  This is just about ready.  I'll approve and merge it once you've merged in the gh-94051 changes.

File: Tools/scripts/generate_global_objects.py
Comment (kumaraditya303) [reply Tools/scripts/generate_global_objects.py:294]: done

File: Tools/scripts/generate_global_objects.py
Comment (kumaraditya303) [reply Tools/scripts/generate_global_objects.py]: done

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

**Comment by ericsnowcurrently** [comment]
Thanks for working on this, @kumaraditya303!

**Comment by kumaraditya303** [comment]
> Thanks for working on this

Thanks for the reviews!

**Pull request merged**
